### ISSUE TITLE
hide sensitive columns in SQL results as well as GUI

### ISSUE THREAD

**Issue by snarfed** [metabase/metabase#1689]
if you mark a column as sensitive in metadata, it no longer appears in GUI-based query results, but still appears in SQL query results. it'd be nice to hide them from SQL too. i know that basically amounts to writing a full scale SQL preprocessor, which is no small feat, and i also know you all are reluctant to do too much with permissions just yet, so just a heads up for tracking.

related to (but different from) #1175 and #1450. cc @youssefm @foofygoofy.

i might also be able to find alternatives that do this withuot metabase code changes. more in a bit.

**Comment by snarfed** [comment]
alternative ideas:
* create views for these tables that don't have the sensitive columns, add them to metabase, and hide the underlying tables
* [set up per-column permissions in postgresql](http://www.postgresql.org/docs/9.4/interactive/sql-grant.html#SQL-GRANT-DESCRIPTION-OBJECTS) that allow the metabase user to select just the non-sensitive columns

**Comment by MrMauricioLeite** [comment]
Another alternative is to deploy a metabase install for each group of your organization. The maintance is a problem here, cause you have to keep all public datamodel updated accross all those metabase instances.

**Comment by borisrorsvort** [comment]
This doesn't seems to be getting enough attention. It's been 5 month. To me it's a no go for any decent size organizations untill this is fixed :wink: .

**Comment by agilliland** [comment]
Hey @borisrorsvort I'm curious to hear a little bit more about your use case and why you feel this is a requirement.  What type of sensitive data are you protecting in your database?  What are the different groups of users you have and why are some users able to see this data and not others?

One of the fundamental questions I would ask regarding this particular proposal is, why are you giving people SQL access to a databases with things in them they shouldn’t have access to?

**Comment by borisrorsvort** [comment]
Just some people have access to different types of data depending on they role on the company. 
Sometime trainees need to access basic data but you wouldnt like them to export you're whole list of users. And especially without some Audit logging.

Also its just inconsistent to allow marking some field as sensitive, preventing making questions out of them and just allowing **anybody** to make a custom sql query with select * from users ad get all the fields your marked as sensitive.

**Comment by snarfed** [comment]
just to follow up, we ended up using per-column permissions to block metabase's db user from reading sensitive columns entirely, like i mentioned originally. it's obviously a very coarse sledgehammer, but it's enough for us right now.

again, writing or integrating a full fledged SQL parser/rewriter is a huge ask. i doubt i'd prioritize it highly if i were you all. one possible alternative is a simple per-user permission for whether users can use raw SQL at all, or maybe one for writing SQL queries and one for viewing them. i expect that's way easier, and might still address most use cases.

**Comment by agilliland** [comment]
Thanks for the feedback guys, this is really useful input.

@borisrorsvort One thing that I guess we haven't done a good job of explaining is the fact that the SQL writing tool is not subject to any of the other metadata choices made in the tool and doesn't even know about them.  So when you are writing SQL it's meant to be as if you were writing it on the db console and the intent is to enable full power to users without any restrictions under the assumption that they know what they are doing.

@snarfed I think your solution sounds pretty solid.  I definitely agree that having the ability to disable the full SQL writing access for some users makes a lot of sense.

**Comment by tlrobinson** [comment]
Perhaps in the short term we should make it clear that "Do Not Include" is about visibility in certain contexts, not strict access control.

I agree enforcing access control using database permissions is the safest route if you truly need to restrict access (also protects against bugs in Metabase, etc)

**Comment by borisrorsvort** [comment]
@agilliland It seems indeed that in the short term we should be able to have different access level for some feature/fields + some ui tips for sensitive data fields

**Comment by salsakran** [comment]
@borisrorsvort we've explicitly tailored the current incarnation of Metabase for small company or group use. We have very limited resources and are focusing our development efforts on features that make data more accessible to more people in a company. We are definitely exploring ways to lock things down but it's not as high a priority compared to getting the tool to a Best-in-Show level in terms of the querying interface.

We would *love* *love* to collect your thoughts on the various knobs you'd like in terms of security, but the fundamental reality is that we're a small team, and are constantly making compromises in what we're able to work on. We will be publishing an external facing roadmap, but in general, unless something is assigned a version number, the core team isn't making any commitment on timing for issues that are submitted. If you're able to sponsor features (or the project) via your employer, we can definitely talk =)

**Comment by borisrorsvort** [comment]
@salsakran I understand completely, thanks for making it clear. Though I doubt that hiding the custom sql interface if you're not an admin is such a big feature.

**Comment by salsakran** [comment]
The problem that comes up is that most often people use the SQL interface to share reports with non-technical users when our GUI isn't able to express the query.

A very common example is a cohort analysis. If we disable the sql interface for non-admins that means that end users can't reuse or edit these shared reports. 

I'd agree that simply adding those ~10 lines of code to check for an admin bit on the toggle would be very simple, but there's a ton of surface area around how people are using our product that would be affected.

**Comment by salsakran** [comment]
Our general recommendation for most people currently is to remove access to sensitive information information in a general Company-Wide metabase instance via views (or in our internal use, by setting up a dedicated data warehouse). 

This is our view of the "right" solution for a place with some engineering bandwidth available on the data infra side. We've had people set up additional data marts with specific data that certain job roles would need access to. It's much much easier to verify data access and partitioning by doing it at the database level than trying to secure differential access permissions on the consumption side. Putting my paranoid tin foil hat on, there will always be bugs, escalations, etc possible ,and you're better off having a hard partition between sensitive and non-sensitive information if you *really* care about the separation. 

Anecdotally, we find the management footprint of a single metabase instance to be super low, and we run multiple instances internally for this exact reason. It allows a crystal clear audit trail, as well as protection against bugs or the more common admin human errors allowing cross role/group data access.

**Comment by salsakran** [comment]
@snarfed really like the notion of sql read/write access bits. definitely something to look at in #1450  + #1175

**Comment by hai-ld** [comment]
Hello,

Not sure whether this issue is the right place to discuss this, but marking a table as **Hidden** or a column as **Do Not Include** has no effect for raw queries. I'm quite surprised to discover this. I think the UI for metadata editing should provide a clear security disclaimer message. Currently, for **Do Not Include**, the message is "Metabase will never retrieve this field. Use this for sensitive or irrelevant information.", which is very misleading.

**Comment by mazameli** [comment]
@hai-ld yeah, you're right: we should make the text in the UI clearer about this. I'll see what I can do. 

If you're trying to keep other users from accessing those sensitive columns via the SQL/native query editor, one thing you can do is restrict access to the SQL editor for their user group in the Admin Panel.

**Comment by jlisyc** [comment]
still happends in 0.32.10

**Comment by williamherry** [comment]
is it possible to use same render component for native sql query result page and explore page, or use the same visibility_type check that explore page use, that way this problem should be fixed,

**Comment by flamber** [comment]
@williamherry Metabase would have to be able to parse and validate the SQL completely, which is not a simple thing to do. There's several other things that would benefit from the functionality, like #10525
The workaround is to limit what Metabase has access to in the database (via privileges or views).
